### ISSUE TITLE
Reintroduce the preresolved dependency optimization

### ISSUE THREAD

**Issue by woodruffw** [pypa/pip-audit#610]
We removed this in #540 for correctness reasons.

However, doing so seems to have caused some bad performance regressions on `brew-pip-audit`'s automation: audits that were previously only bound by vulnerability API lookup time are now much slower, as they call into `pip` and do dependency machinery work.

We should consider reintroducing a variant of these, potentially stuffed behind a scary flag.

CC @tetsuo-cpp for thoughts.

CC @alex for visiblity.

**Comment by pauloxnet** [comment]
In our projects, we fixed pip-audit to version 2.5.2 because the later versions are definitely so much slower.

### Version 2.5.2
```console
$ time python3 -m pip_audit --require-hashes --requirement requirements/remote.txt
No known vulnerabilities found

real    0m0,624s
user    0m0,563s
sys     0m0,062s
```

### Version 2.5.5
```console
$ time python3 -m pip_audit --require-hashes --requirement requirements/remote.txt
No known vulnerabilities found

real    0m15,915s
user    0m11,808s
sys     0m1,037s
```

**Comment by tetsuo-cpp** [comment]
@pauloxnet
Thanks for letting us know!

Just curious, is your requirements file generated by `pip-compile`?

**Comment by tetsuo-cpp** [comment]
@woodruffw I wonder whether this would be solved by us passing `--no-deps` into `pip` whenever the user gives us `--require-hashes`. One of the key differences in `pip`'s behaviour is that `--require-hashes` doesn't imply `--no-deps` and it will do full dependency resolution to ensure that all requirements are in the file and are hashed.

I need to do a bit of testing with `pip`, but the solution might be to just pass in `--no-deps` whenever the user gives us `--require-hashes` and hopefully this stops dependency resolution from running.

**Comment by tetsuo-cpp** [comment]
@pauloxnet If you run that command with `--no-deps`, does that help performance at all?

**Comment by pauloxnet** [comment]
I confirm that we create hashed requirements file with `pipcompile`

```console
python3 -m piptools compile --generate-hashes --no-header --quiet --resolver=backtracking --upgrade --output-file requirements/test.txt requirements/test.in
```

Using `--no-deps` does not help performances.

The main difference in `pip-audit` up to 2.5.2 seems to me that it checks the versions only considering the requirements file, while in later versions it tries to install them in an isolated environment, but this last thing is slower and potentially forces you to carry around more addictions.

**Comment by tetsuo-cpp** [comment]
@woodruffw Based on that, we might have to just bring that logic back. I don't think it's that big a deal as long as it's not the default behaviour when someone uses `--require-hashes`. It only has problems with local requirements and even then, the offending requirements are listed in the "skipped" summary.

**Comment by woodruffw** [comment]
@tetsuo-cpp Agreed; I'll update this issue.

**Comment by trottomv** [comment]
Hi everyone, I confirm what @pauloxnet reported and the drop in performance from versions after 2.5.2, in the event that a requirement.txt file built with pip compile and hashes is fed to pip audit, due to the fact that from version 2.5.3 all the dependencies parsed by pip audit are installed in a temporary virtualenv, with no possibility to exclude this option.

While introspecting the pip audit code, I came to this reflection, that it is logically correct to force the installation of packages on a temporary virtual env to make sure that the audit is also performed on all subdependencies. The only, or almost the only case, in which this option could be skipped is in the case where the requirements are compiled with pip compile, and not necessarily with the hashes, as there are python libraries that generate the hashes on the requirement.txt without the assurance that these contain all the subdependencies.
I therefore believe that, in general, it is correct that while passing the arguments --require-hashes or --no-deps the installation takes place on the temporary virtualenv, because the presence of the hashes does not give the certainty that all the dependencies present on the requirements.txt are all those will actually be installed.

Personally I would suggest setting up a new code flow to provide pip audit with information about whether the requirements.txt was compiled with pip compile or not. Maybe a new pip audit run args? e.g. `pip audit --skip-venv` or `pip audit --requirement requirements.txt --with-pip-compile`

Alternatively, it could also try to parse the content of the requirements.txt which, if it contains `pip-compile` in the header, could trigger the most performing flow that avoids installing the dependencies in the temporary virtualenv. But bear in mind that `pip compile` can be run with the `--no-header` option, at which point even though the requirement was compiled with `pip compile` it would not lead to a more performing flow.

**Comment by trottomv** [comment]
@woodruffw I thought of opening a pull request with the proposal to introduce a new option to optimize dependency resolution in case they are compiled with pip-compile ☝️

**Comment by woodruffw** [comment]
The work in #626 has been released with 2.6.0. Thanks for your hard work here @trottomv!

**Comment by pauloxnet** [comment]
Thanks to @woodruffw and @trottomv for the new version.
